site/app/controllers/grading/ElectronicGraderController.php

```diff
@@ -2339,6 +2339,84 @@ public function saveGradedComponent(TaGradedGradeable $ta_graded_gradeable, Grad
         }
     }
 
+    /**
+     * Route for saving the marks the submitter received for a component
+     */
+    #[Route("/courses/{_semester}/{_course}/gradeable/{gradeable_id}/grading/graded_gradeable/change_grade_version", methods: ["POST"])]
+    public function ajaxUpdateGradedVersionForStudent(string $gradeable_id): void {
+        $anon_id = $_POST['anon_id'] ?? null;
+        $graded_version = intval($_POST['graded_version'] ?? null);
+        $component_ids = $_POST['component_ids'] ?? [];
+
+        if ($anon_id === null) {
+            $this->core->getOutput()->renderJsonFail('Missing anon_id parameter');
+            return;
+        }
+        if ($graded_version < 1) {
+            $this->core->getOutput()->renderJsonFail('Invalid graded_version parameter');
+            return;
+        }
+
+        if (count($component_ids) < 1) {
+            $this->core->getOutput()->renderJsonFail('Missing component_ids parameter');
+            return;
+        }
+
+        // Get the gradeable
+        $gradeable = $this->tryGetGradeable($gradeable_id);
+        if ($gradeable === false) {
+            return;
+        }
+        // Get user id from the anon id
+        $submitter_id = $this->tryGetSubmitterIdFromAnonId($anon_id, $gradeable_id);
+        if ($submitter_id === false) {
+            return;
+        }
+
+        // Get the graded gradeable
+        $graded_gradeable = $this->tryGetGradedGradeable($gradeable, $submitter_id);
+        if ($graded_gradeable === false) {
+            return;
+        }
+
+
+        // check if all the components provided are real and the user has permission
+        foreach ($component_ids as &$component_id) {
+            $component = $this->tryGetComponent($gradeable, $component_id);
+            if ($component === false) {
+                $this->core->getOutput()->renderJsonFail("Invalid component id \"$component_id\"");
+                return;
+            }
+            elseif (!$this->core->getAccess()->canI("grading.electronic.save_graded_component", ["gradeable" => $gradeable, "graded_gradeable" => $graded_gradeable, "component" => $component])) {
+                $this->core->getOutput()->renderJsonFail("Insufficient permissions to change graded version of component $component_id");
+                return;
+            }
+        }
+
+        $logger_params = [
+            "course_semester" => $this->core->getConfig()->getTerm(),
+            "course_name" => $this->core->getDisplayedCourseName(),
+            "gradeable_id" => $gradeable_id,
+            "grader_id" => $this->core->getUser()->getId(),
+            "component_ids" => implode(",", $component_ids),
+            "action" => "CHANGE_GRADEABLE_GRADED_VERSION",
+            "submitter_id" => $submitter_id
+        ];
+        Logger::logTAGrading($logger_params);
+
+        try {
+            // do thing
```

Comment:
do thing

Reply:
this comment should either be removed or improved.